src/shims/fs.rs

```diff
@@ -257,6 +263,214 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         this.try_unwrap_io_result(result)
     }
 
+    fn statx(
+        &mut self,
+        dirfd_op: OpTy<'tcx, Tag>,
+        pathname_op: OpTy<'tcx, Tag>,
+        flags_op: OpTy<'tcx, Tag>,
+        _mask_op: OpTy<'tcx, Tag>,
+        statxbuf_op: OpTy<'tcx, Tag>
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("statx")?;
+
+        let statxbuf_scalar = this.read_scalar(statxbuf_op)?.not_undef()?;
+        let pathname_scalar = this.read_scalar(pathname_op)?.not_undef()?;
+
+        // If the statxbuf or pathname pointers are null, the function fails with `EFAULT`.
+        if this.is_null(statxbuf_scalar)? || this.is_null(pathname_scalar)? {
+            let efault = this.eval_libc("EFAULT")?;
+            this.set_last_error(efault)?;
+            return Ok(-1);
+        }
+
+        // If statxbuf is not null we need to produce a proper `MemPlace` to write the results of
+        // the function to it. This is needed because all the `statx` parameters are provided with
+        // an `i64` layout instead of the proper type layout.
+        let statxbuf_ptr = this.force_ptr(statxbuf_scalar)?;
+        // This long path is required instead of `libc::statx` because the `statx` struct and
+        // function share the same namespace.
+        let statx_ty = this
+            .resolve_path(&["libc", "unix", "linux_like", "linux", "gnu", "statx"])?
```

Comment:
feels a bit fragile to rely on private libc paths (or is this public?)

Reply:
well I don't really know how to address this one. The problem is that `libc::statx` points both to the `statx` function and the `statx` struct but `resolve_path` gets the function instead.

Reply:
You could split out https://github.com/rust-lang/miri/blob/00c8019720a12f6a539f59071c15d2bdeec1f9fe/src/helpers.rs#L22-L47 into a separate function that returns an iterator and use `flat_map` instead of `and_then` and make https://github.com/rust-lang/miri/blob/00c8019720a12f6a539f59071c15d2bdeec1f9fe/src/helpers.rs#L35 not be a for loop but an iterator that is returned. This may get a bit ugly and I can't come up with a way on the spot, but it should be possible without boxing. Worst case you have to create a custom iterator type that contains the `item` field.

Then you get a list of all items, and can check whether one of the items is in fact a struct and not a function

Reply:
An iterator seems overkill... and also not a great API. I think we should have `resolve_type_path` and `resolve_X_path` to resolve paths in the two namespaces (types and... how do we call the other things? functions and consts and statics and all of that?).

Reply:
That's the value namespace.

Reply:
I think that we would need to make https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.guess_def_namespace public then.

Reply:
`resolve_value_path` then.

> I think that we would need to make https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.guess_def_namespace public then.

"guess" doesn't sound very promising though...
Feel free to leave a FIXME; this doesn't have to be sorted in this PR if the public API of rustc  currently is too limited. But I feel if we have to namespaces (well, three, with macros), we should fairly simply have two methods for doing lookups in them.

Reply:
I tried to use `DefKind::matches_ns` too but it didn't work. I think it might be because `tcx.def_kind`returns `Option<DefKind>`. I wonder if `statx` doesn't have a proper `DefKind`.

Reply:
@christianpoveda  Please don't resolve conversations that we haven't actually resolved.

So, for this one... we should at least open an issue to track this somewhere. We definitely want a reliable way to look up names in the type vs the value namespace.

Reply:
should I open the issue in the miri repo or in the rust repo?

Reply:
Miri repo seems fine.

Reply:
Opened at https://github.com/rust-lang/miri/issues/1123